### PR TITLE
Pyverbs fork tests

### PR DESCRIPTION
**Pull request by galpress**
This PR adds basic fork unit testing by calling ibv_fork_init() and ibv_is_fork_initialized().
The fork tests are impacted by the order of tests execution, for example, if a previous test has registered an MR then ibv_fork_init() is expected to fail. If only the fork tests are run, it should succeed.

### CODE REVIEW

**Comment by rleon** [comment]
> The fork tests are impacted by the order of tests execution,

This doesn't sound good.

Thanks

**Comment by galpress** [comment]
To clarify, they obviously always pass regardless of the order of execution.
But it means that the test has to take into account that ibv_fork_init() could fail for example.

I'm open for suggestions if anyone has an elegant solution.

**Comment by galpress** [comment]
@EdwardSro thoughts?

File: tests/test_fork.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            fork_init()
+        except PyverbsRDMAError as ex:
+            # Depends on the order of the tests EINVAL could be returned if
+            # fork_init() is called after an MR has already been registered.
+            self.assertEqual(ex.error_code, errno.EINVAL)
+
+    def test_is_fork_initialized(self):
+        try:
+            fork_init()
+            expected_ret = [e.IBV_FORK_ENABLED, e.IBV_FORK_UNNEEDED]
```
Comment (EdwardSro) [inline tests/test_fork.py:31]: you can call is_fork_initialized() before calling fork_init and check its returned value to make a more "precise" check after fork_init.

File: tests/test_fork.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from tests.base import PyverbsAPITestCase
+
+
+class ForkAPITest(PyverbsAPITestCase):
+    """
+    Test the API of the fork functions.
+    """
+
+    def setUp(self):
+        super().setUp()
```
Comment (EdwardSro) [inline tests/test_fork.py:18]: no need, the parent's setup would be called automatically...

File: pyverbs/libibverbs.pxd [hunk 3]
```diff
@@ -4,7 +4,7 @@
 include 'libibverbs_enums.pxd'
 from libc.stdint cimport uint8_t, uint16_t, uint32_t, uint64_t
 from posix.time cimport timespec
-
+cimport pyverbs.libibverbs_enums as e
```
Comment (EdwardSro) [inline pyverbs/libibverbs.pxd:7]: no need, there's an "include 'libibverbs_enums.pxd'" line, so you can directly use the enums

File: pyverbs/libibverbs.pxd [hunk 4]
```diff
@@ -723,7 +723,8 @@ cdef extern from 'infiniband/verbs.h':
     int ibv_query_rt_values_ex(ibv_context *context, ibv_values_ex *values)
     int ibv_get_async_event(ibv_context *context, ibv_async_event *event)
     void ibv_ack_async_event(ibv_async_event *event)
-
+    int ibv_fork_init()
+    e.ibv_fork_status ibv_is_fork_initialized()
```
Comment (EdwardSro) [inline pyverbs/libibverbs.pxd:727]: remove the "e." after removing the enums cimport

**Review by EdwardSro — CHANGES_REQUESTED** [review]
If memory registration was done, fork_init would fail? regardless whether the memory was freed by the user or not?

**Comment by galpress** [comment]
Yes, if ibv_reg_mr() is called prior to ibv_fork_init() then the call fails:
https://github.com/linux-rdma/rdma-core/blob/master/libibverbs/man/ibv_fork_init.3.md#return-value

File: tests/test_fork.py (hunk 1 above)
Comment (galpress) [reply tests/test_fork.py:31]: I didn't want to get into corner cases with ordering of the different test cases.
If a prior test case already called fork_init() then is_fork_initialized() could return IBV_FORK_ENABLED, even if fork_init() is not called yet in this test case.

File: tests/test_fork.py (hunk 2 above)
Comment (galpress) [reply tests/test_fork.py:18]: Done

File: pyverbs/libibverbs.pxd (hunk 3 above)
Comment (galpress) [reply pyverbs/libibverbs.pxd:7]: Thanks, didn't know that's how it works.

File: pyverbs/libibverbs.pxd (hunk 4 above)
Comment (galpress) [reply pyverbs/libibverbs.pxd:727]: Done

File: tests/test_fork.py (hunk 1 above)
Comment (EdwardSro) [reply tests/test_fork.py:28]: I know, what I meant, is to call is_fork_initialized(), according to the returned value you would know the exact expected value of the second is_fork_initialized call after fork_init().
But yes, then you're required to handle multiple cases.
You can keep it as is... it wouldn't have much added value.

**Review by EdwardSro — COMMENTED** [review]
Are you planning to use these verbs in future pyverbs tests?
Since the tests don't have much-added value (still they're needed to cover newly added verb(s) to pyverbs) IMO, I would drop the first test (test_fork_init) and keep the second one. What do you think?

**Comment by galpress** [comment]
> Are you planning to use these verbs in future pyverbs tests?

I have some thoughts about using it in other tests, but I don't think that would happen soon.

> IMO, I would drop the first test (test_fork_init) and keep the second one. What do you think?

Sure, will do.

Thanks for the review!

**Review by EdwardSro — APPROVED** [review]
Thanks.
